src/dump.c

```diff
@@ -3100,14 +3126,19 @@ static jl_value_t *_jl_restore_incremental(ios_t *f)
     // now all of the interconnects will be created
     jl_recache_types(); // make all of the types identities correct
     init_order = jl_finalize_deserializer(&s, tracee_list); // done with f and s (needs to be after recache types)
+
+    jl_array_t *extmeths = jl_alloc_vec_any(0);  // make array of references for GC
+    linkedlist_to_array(extmeths, &external_methods);
+
+    JL_GC_PUSH3(&init_order, &restored, &extmeths);
+    jl_gc_enable(en);
+
     jl_insert_methods(&external_methods); // hook up methods of external generic functions (needs to be after recache types)
     jl_recache_other(); // make all of the other objects identities correct (needs to be after insert methods)
```

Comment:
After this step, some new objects (from the flagref list) may be only accessible from old objects. I think we'll either need to update the flagref to track the parent so we can add the gc_wb in here, or prevent promotion.